specifications/trusted_computing_base/safety_rules.md

```diff
@@ -0,0 +1,218 @@
+* A previous epoch is impossible otherwise this would violate the `waypoint` verification performed earlier.
+
+**Notes**:
+
+1. If LSR finds that it is unable to find its expected `consensus_key` in storage it will error.
+2. LSR depends on storage to have been initialized with `epoch`, `last_voted_round`, `preferred_round`, and `waypoint`. At Genesis, these are set to `0`,`0`, `0`, and the `ledger_info` after Genesis.
+3. LSR can leverage the `validator_verifier` to identify the current expected `consensus_key` and **must** not participate in any other operation unless LSR has access to this key.
+
+### Voting -- `construct_and_sign_vote`
+
+Consensus relies on LSR to sign votes. In order to ensure these properties, Consensus provides a `VoteProposal` that may have been signed by LEC if enabled.  LSR verifies and updates state via the properties mentioned above (1, 2, 3, 5), then signs a `Vote` derived from the `VoteProposal` and returns it to consensus.
+
+Prior to ensuring any properties, LSR must first check if LSR has a valid `consensus_key` for signing the proposal and that the proposal's `epoch` matches the current `epoch`.
+
+#### Property 1
+
+> A validator can only vote on proposals newer than those they have already voted on as defined by linearly increasing epochs and rounds.
+
+LSR enforces this by keeping track of `last_voted_round`, this is updated immediately prior to LSR voting on a proposal. Therefore LSR can never vote on the same round twice or an earlier round than the current.
+
+#### Property 2
+
+> A validator can only vote on a proposal if it extends from a proposal that a quorum of voters have seen.
+
+LSR enforces this by ensuring that the `quorum_cert` associated with a proposal is properly constructed, is correctly signed leveraging the `validator_verifier`, and that the parent's round within the `quorum_cert` is equal to or greater than the `preferred_round`.  The certified state of parent round within the `quorum_cert` is guaranteed to have been seen by a quorum of validators and hence serves as a bookmark to not backtrack to earlier states.  Because of consensus 3-chain commit rule, violations of this could break safety.
+
+LSR updates the persistent storage `preferred_round` to the maximum of the `quorum_cert`'s parent's round or the current stored `preferred_round`.
+
+#### Property 3
+
+> A validator can only vote on a proposal if the output of that proposal extends its parent's output.
+
+Each `Vote` contains the executed state output, the resulting root hash of the blockchain ledger after executing the proposal. In order to verify that this extends the parent block, a `VoteProposal` contains an `AccumulatorExtensionProof` which is a merkle-proof showing how the current state can be derived from the parent's state, hence ensuring that the ledger is append-only.
+
+#### Property 5
```

Comment:
a quick glance tells me that there is a missing prop 4 here

Reply:
weird, did you just copy and paste this @JoshLind ?

Reply:
Yeah, I literally just copied it across. Seems like a problem with the original: https://github.com/novifinancial/libra-specs/blob/master/trusted_computing_base/safety_rules.md